lib/core/network/protocols/http+websocket.js

```diff
@@ -0,0 +1,920 @@
+        socket.count = 1;
+      }
+    }
+
+    let parsed;
+    const message = Buffer.from(data).toString();
+
+    debugWS('[%s] client message: %s', connection.id, message);
+
+    try {
+      parsed = JSON.parse(message);
+    }
+    catch (e) {
+      /*
+       we cannot add a "room" information since we need to extract
+       a request ID from the incoming data, which is apparently
+       not a valid JSON
+       So... the error is forwarded to the client, hoping they know
+       what to do with it.
+       */
+      this.wsSendError(
+        socket,
+        connection,
+        kerrorWS.getFrom(e, 'unexpected_error', e.message));
+      return;
+    }
+
+    if (parsed.p && parsed.p === 1 && Object.keys(parsed).length === 1) {
+      debugWS('[%s] sending back a "pong" message', connection.id);
+      this.wsSend(socket, WS_APP_PONG_RESPONSE);
+      return;
+    }
+
+    try {
+      this.entryPoint.execute(new Request(parsed, { connection }), result => {
+        if (result.content && typeof result.content === 'object') {
+          result.content.room = result.requestId;
+        }
+        this.wsSend(socket, Buffer.from(JSON.stringify(result.content)));
+      });
+    }
+    catch (e) {
+      const errobj = {
+        error: {
+          message: e.message
+        },
+        room: parsed.requestId,
+        status: 400
+      };
+
+      this.wsSend(socket, Buffer.from(JSON.stringify(errobj)));
+    }
+  }
+
+  /**
+   * Absorb as much of the backpressure buffer as possible
+   * @param  {uWS.WebSocket} socket
+   */
+  wsOnDrainHandler (socket) {
+    socket.cork(() => {
+      const buffer = this.backpressureBuffer.get(socket);
+
+      while (buffer.length > 0
+        && socket.getBufferedAmount() < WS_MAX_BACKPRESSURE
+      ) {
+        const payload = buffer.shift();
+        socket.send(payload);
+      }
+    });
+  }
+
+  /**
+   * Forwards an error to a socket
+   *
+   * @param  {uWS.WebSocket} socket
+   * @param  {ClientConnection} connection
+   * @param  {Error} error
+   */
+  wsSendError (socket, connection, error) {
+    const request = new Request({}, { connection, error });
+    const sanitized = removeErrorStack(request.response.toJSON()).content;
+
+    this.wsSend(socket, Buffer.from(JSON.stringify(sanitized)));
+  }
+
+  /**
+   * Sends a message immediately, or queue it up for later if backpressure built
+   * up
+   *
+   * @param  {uWS.WebSocket} socket
+   * @param  {Buffer} payload
+   */
+  wsSend (socket, payload) {
+    if (!this.connectionBySocket.has(socket)) {
+      return;
+    }
+
+    if (socket.getBufferedAmount() < WS_MAX_BACKPRESSURE) {
+      socket.cork(() => socket.send(payload));
+    }
+    else {
+      const buffer = this.backpressureBuffer.get(socket);
+      buffer.push(payload);
+
+      // Client socket too slow: we need to close it
+      if (buffer.length > WS_BACKPRESSURE_BUFFER_MAX_LENGTH) {
+        socket.end(1011, WS_BACKPRESSURE_MESSAGE);
```

Comment:
What is this number `1011` ?

Reply:
Websocket termination code, see the rfc: https://tools.ietf.org/html/rfc6455#page-46

Reply:
Can we have a named constant instead of a number? It's more readable than having to read the RFC :sweat_smile:

Reply:
Good point :+1: